content/guides/getstarted/rust-to-solana.md

```diff
@@ -0,0 +1,186 @@
+
+        //program code goes here
+
+    }
+```
+These three parameters are passed to every on-chain program. 
+1. The `program_id` is the public key of the current program. 
+2. The `accounts` are all accounts that are required to process the instruction. 
+3. The `instruction_data` is data specific to that instruction.
+
+Every program must then call the `entrypoint!` macro on the instruction, as
+follows:
+
+```rust
+entrypoint!(process_instruction);
+```
+
+### Building and Testing 
+After installing the Solana command-line tools, projects can be built to target host machines as normal with `cargo build`. 
+
+However, to target the Solana runtime, use `cargo build-bpf` or `cargo build-spf` which will compile the program to the bytecode necessary to run it on the Solana runtime.
+
+Unit testing can be achieved via `cargo test` with standard `#test` attributes. For more integrated testing, the [solana-program-test](https://crates.io/crates/solana-program-test) crate provides a local Solana runtime instance which can be used in conjunction with outside tests sending transactions.
+
+Finally, a full test cluster can be started with the [solana-test-validator](https://docs.solanalabs.com/cli/examples/test-validator), installed along with the Solana CLI. This creates a fully featured test cluster on a local machine, which can then deploy programs and run tests.
+
+## Understanding Restrictions 
+While most standard Rust crates are available in the Solana runtime, and third-party crates are supported as well, there are several limitations. Since the Solana runtime has resource constraints and must run deterministically, here are the differences to be aware of:
+
+### Package Limitations 
+The following packages are unavailable:
+rand
+* std::fs
+* std::net
+* std::future
+* std::process
+* std::sync
+* std::task
+* std::thread
+* std::time
+
+The following packages have limited functionality:
+* std::hash
+* std::os
+
+### Rand Dependencies 
+Because programs must run deterministically, the `rand` crate is not available. Using an additional crate that depends on `rand` will also cause compile errors. 
+
+However, if the crate used simply depends on `rand` but does not actually generate random numbers, then it is possible to work around this by adding the following to the program’s Cargo.toml:
+
+```toml
+[dependencies]
+getrandom = { version = "0.1.14", features = ["dummy"] }
+```
+
+### Macro Changes
+Several standard macros have been replaced or have modified behavior to be aware of. First, the `println!` has been replaced with the computationally simpler `msg!` macro. The `msg!` macro outputs to the program logs, and it can be used as follows:
+
+```rust 
+msg!(“Your message”);
+msg!(0_64, 1_64, 2_64);
+msg!(“Your variable: {:?}”, variable);
+```
+
+The `panic!`, `assert!`, and any internal panics are also output to the program logs by default. However, this can be modified with a custom panic handler.
+
```

Comment:
Maybe also worth mentioning Error handling as a better alternative to panic? 

smth like this? 
#[error_code]
pub enum MyErrors {
    CustomError,
}

#[program]
pub mod anchor_error_test {
    use super::*;

    pub fn initialize(ctx: Context<Initialize>) -> Result<()> {
        //panic!("PANIC");
        return Err(MyErrors::CustomError.into());

        Ok(())
    }
}